### DISCUSSION TITLE
How to best integrate DaisyUI in multilanguage application?

### DISCUSSION THREAD

**Discussion by VesaLahd** [saadeghi/daisyui#1928] [Q&A] [answered]
As the title stated. Building a multilanguage application with internationalization needs. Seems that the RTL configuration is just buildtime flag for enabling the components to be exported from RTL implementations. However this seems like a hindrance when implementing runtime language changes in the application itself since the reading direction might change between languages. 

Is there some proper way to support this kind of functionality using DaisyUI?

**Comment by saadeghi** [comment, marked as answer]
You should use 2 separate CSS files and here's the reason:  
It's not a good idea to use 1 CSS file for both directions because for RTL, every directional CSS property (left, right, margin, padding, etc) should switch direction on both daisyUI class names and your Tailwind CSS utility classes.  
If we wanted to provide a bidirectional style for daisyUI's CSS we had to wrap all of them in `[dir="ltr"]/[dir="rtl"]`
- it would increase the file size unnecessarily
- it would mess with CSS specificity so if the `.btn` style was `[dir="ltr"] .btn"` and it has a left margin you couldn't simply use a Tailwind CSS utility class to customize it because the original style has a higher CSS specificity.
- it would be just daisyUI classes and you should still flip your Tailwind CSS utility classes somehow and there would be no point to do this separately because after all, daisyUI class names are getting processed by Tailwind CSS anyway.
- There is a high chance, users only pick the language once and they use the site with the chosen language forever. So if a user picked a RTL language there's no reason to send the `ltr` styles to their browsers everytime. It's unused CSS for them.

So the most efficient way is to build 2 files separately. One for LTR and one for RTL and send the correct file to the browser when user changes the language.

And here's an example on how to do that using 2 Tailwind configs:  
https://github.com/saadeghi/daisyui-bidirectional-builds

The repo is a little old but you can see how it works.


Let me know if you have any questions

**Reply by VesaLahd** [reply]
Yeah, figured that this might be the case. Thanks for the quick answer.

**Reply by mirismaili** [reply]
@saadeghi 

A lot of CSS properties are direction-aware now. There are a few cases that need different css-adjustment for RTL/LTR directions. We can't completely build 2 different applications, just for direction! Then how do we handle routing? Out of Next.js?

**Reply by mirismaili** [reply]
And why this is LTR:

https://daisyui.com/?lang=ar

![image](https://github.com/saadeghi/daisyui/assets/15800189/5ef3566d-bd10-40ca-8746-63fe91e94ba3)

@saadeghi

**Reply by saadeghi** [reply]
Tomorrow daisyUI 4 will be released with automatic RTL in runtime. Thanks for your patience 🫡
